server/src/test/java/org/apache/druid/segment/loading/OmniDataSegmentArchiverTest.java

```diff
@@ -62,6 +64,21 @@ public void testArchiveSegmentUnknowType()
     );
   }
 
+  @Test
+  public void testBadSegmentArchiverAccessException()
+  {
+    final DataSegment segment = Mockito.mock(DataSegment.class);
+    Mockito.when(segment.getLoadSpec()).thenReturn(ImmutableMap.of("type", "bad"));
+
+    final Injector injector = createInjector(null);
+    final OmniDataSegmentArchiver segmentArchiver = injector.getInstance(OmniDataSegmentArchiver.class);
+    Assert.assertThrows(
```

Comment:
I suppose we are doing this to ensure that the instantiation of the `BadSegmentArchiver` happens lazily, i.e. only when `archive` is called and not when `OmniDataSegmentArchiver` is constructed.

Throwing an exception to verify this is probably not ideal as it gives the wrong impression while reading the test.

I would advise using something like an `AtomicInteger instanceCounter` and verifying its value before and after the invocation of `segmentArchiver.archive` in the test. We could either
- inject the `instanceCounter` to the `BadSegmentArchiver` constructor (preferable)
- or, use a static `instanceCounter` field in the test itself.

this could even help in verifying the memoization.

The same thing could be done for other tests too.

Reply:
though having the exception helps ensure that an out-of-turn initialization will fail the test. is the current approach not simpler? The class does have `Bad` in its name.

Reply:
The test will still fail if the instanceCount is not zero before we call `segmentArchiver.archive`.
So what I am suggesting would look like this:

```
assertEquals(instanceCount, 0);
segmentArchiver.archive();
assertEquals(instanceCount, 1);

// verify memoization
segmentArchiver.archive();
assertEquals(instanceCount, 1);
```

Since this test is really trying to verify that instantiation happens at the right invocation, having an `assertThrows` in there feels weird. I would normally use `assertThrows` when I am trying to verify
handling of a known bad argument or bad state and an exception which the code actually throws.

Reply:
it's not the `archive` call that we are testing but the lazy instantiation. The memoization tests are already there in `OssStorageDruidModuleTest` which is why I didn't add them here again.  There are similar memoization tests in a few other modules. 

I feel that style is of course going to vary from person to person. It's also the test code. And we have enough coverage for the tests. Readability is of course very subjective. I like to have as less global variables as possible. To me, those make reading the code difficult. The current structure seems simpler. There is a bad class and initializing it is going to result in a failure.

Reply:
Sounds good.